support gpu direct rdma

randomkang opened the pull request:
### What problem does this PR solve?

Issue Number: resolve #3102

Problem Summary:

### What is changed and the side effects?

Changed: 

1) recv all data on gpu first
2) the gpu block is alloced from a gpu block pool
3) brpc header, meta and body will be copied from gpu to cpu to process. 
4) To decrease the d2h counts, we will prefetch 512B to memory

Side effects:
- Performance effects:

- Breaking backward compatibility: 

---
### Check List:
- Please make sure your changes are compilable.
- When providing us with a new feature, it is best to add related tests.
- Please follow [Contributor Covenant Code of Conduct](https://github.com/apache/brpc/blob/master/CODE_OF_CONDUCT.md).

yanglimingcn commented on src/brpc/rdma/rdma_endpoint.cpp:
```diff
@@ -922,31 +933,57 @@ int RdmaEndpoint::SendImm(uint32_t imm) {
 
 ssize_t RdmaEndpoint::HandleCompletion(ibv_wc& wc) {
     bool zerocopy = FLAGS_rdma_recv_zerocopy;
+    //LOG(INFO) << "Handle Completion: " << wc.opcode;
```
This part has a new PR #3145

randomkang replied:
This solution with two windows has been tested in our tasks and it works well.

randomkang commented on src/brpc/policy/baidu_rpc_protocol.cpp:
```diff
@@ -101,44 +106,116 @@ static void SerializeRpcHeaderAndMeta(
+
     char header_buf[12];
-    const size_t n = source->copy_to(header_buf, sizeof(header_buf));
+    size_t n = 0;
```
Maybe put GPU functionality into a separate file or define a new protocol is two heavy, we reorg the code to decrease the macro definitions. Is it ok?

zchuang185 replied:
I think it's better to separate new protocol file for GDR feature， Besides, I have a idea we can also define a new protocol meta that includes both request meta and response meta carrying the GDR attribute value., the corresponding meta should also contain the GPU address information that both sides need to register with IB. Doing the GDR operations inside processRequest and ProcessRpcResponse method step will probably be more reasonable.

chenBright commented on src/brpc/rdma/rdma_endpoint.cpp:
```diff
@@ -1032,11 +1070,20 @@ int RdmaEndpoint::PostRecv(uint32_t num, bool zerocopy) {
         if (DoPostRecv(_rbuf_data[_rq_received], g_rdma_recv_block_size) < 0) {
             _rbuf[_rq_received].clear();
             return -1;
         }
```
There are two posts here, which should be the reason why #3145 with GDR has too many IMMs.

randomkang committed:
support gpu direct rdma
1) recv all data on gpu first
2) the gpu block is alloced from a gpu block pool
3) brpc header, meta and body will be copied from gpu to cpu to process.
4) To decrease the d2h counts, we will prefetch 512B to memory

Co-authored-by: sunce4t <847480001@qq.com>

randomkang committed:
review by deepseek

randomkang committed:
review by glm5.2

randomkang committed:
fix build error

randomkang commented:
@yanglimingcn @zchuango all comments have been resolved, please review again.

yanglimingcn commented on docs/cn/gdr.md:
```diff
@@ -0,0 +1,43 @@
+GDR新增了一个显存池，类似于RDMA内存池，显存池的数据也是按照block进行组织的。
+当打开GDR功能后，框架通过DoPostRecvGDR来发起显存上的WQE。
+在接收到数据后，我们将header、meta、body（不包括attachment）copy回内存进行处理。
+AttachMent位于显存上，用户可以调用IOBuf::copy_from_gpu接口将attachment从brpc框架层copy到应用层进行处理。
```
Attachment

yanglimingcn commented on docs/cn/gdr.md:
```diff
@@ -0,0 +1,43 @@
+
+
+注意：
+1. 在使用gdr功能时，需要将环境变量MLX5_SCATTER_TO_CQE设置为0.
```
这个能说明一下原因吗？

yanglimingcn commented on src/brpc/acceptor.h:
```diff
@@ -111,7 +111,7 @@ friend class Server;
     std::shared_ptr<SocketSSLContext> _ssl_ctx;
 
-    // Choose to use a certain socket: 0 TCP, 1 RDMA
+    // Choose to use a certain socket: 0 TCP, 1 RDMA, 2 GDR
```
GDR mode is 3

yanglimingcn commented on src/butil/gpu/gpu_block_pool.cpp:
```diff
@@ -0,0 +1,507 @@
+  static int original_device = -1;
+
+  int current_device;
+  cudaGetDevice(&current_device);
```
Should we need to check return value

yanglimingcn commented:
Turn multiple commits into one

chenBright commented on src/butil/iobuf.h:
```diff
@@ -146,6 +146,17 @@ friend class SingleIOBuf;
     size_t cutn(void* out, size_t n);
     size_t cutn(std::string* out, size_t n);
+
+#if BRPC_WITH_GDR
```
支持 gpu 内存后，IOBuf 似乎就没法安全使用了，很多接口都是不能用的。

chenBright commented on src/brpc/policy/baidu_rpc_protocol_gpu.cpp:
```diff
@@ -0,0 +1,237 @@
+        return MakeParseError(PARSE_ERROR_TRY_OTHERS);
+    }
+    butil::gdr::BlockPoolAllocator* host_allocator = butil::gdr::BlockPoolAllocators::singleton()->get_cpu_allocator();
+    prefetch_d2h_data = host_allocator->AllocateRaw(prefetch_d2h_size);
```
这里的拷贝似乎并不是一个好的设计。异构的内存，最好能通过不通的数据通道传输。

randomkang commented:
> Turn multiple commits into one

ok

yanglimingcn commented:
LGTM

yanglimingcn commented on src/butil/gpu/gpu_block_pool.cpp:
```diff
@@ -0,0 +1,509 @@
+    return (void *)d_data;
+}
+
+void* get_cpu_mem(int gpu_id, int64_t cpu_mem_size) {
```
这个入参，应该是gpu_id吗

randomkang replied:
是的。这里申请的是pinned memory，使用pinned memory和显存进行交换，会比普通的memory快。在使用cudaMallocHost分配pinned memory的时候，建议先执行cudaSetDevice，这样分配的pined memory就会与对应的gpu id关联。

yanglimingcn commented on src/butil/gpu/gpu_block_pool.cpp:
```diff
@@ -0,0 +1,509 @@
+
+std::atomic<BlockPoolAllocators*> BlockPoolAllocators::instance_{nullptr};
+
+BlockPoolAllocators* BlockPoolAllocators::singleton() {
```
你看brpc自带的singleton.h是否能满足你的需求。

randomkang replied:
我看了下，可以用brpc自带的Singleton，已改

yanglimingcn commented on src/butil/iobuf.cpp:
```diff
@@ -1201,6 +1275,15 @@ uint64_t IOBuf::get_first_data_meta() {
 
+void* IOBuf::get_first_data_ptr() {
+    if (_ref_num() == 0) {
+        return 0;
```
这里返回nullptr

randomkang replied:
OK
